### PR TITLE
Make clipping configurable. Closes #2536

### LINKED ISSUE #2536: Feature request: make clipping configurable
I frequently would like to be able to draw things that fall outside of the plot panel. This is currently not possible, as clipping is strictly enforced and not configurable. I think it would be nice if this could be switched on or off. First, as motivation, an example where it would be useful. Then my proposal of how it could be implemented.

Consider this plot:
```{r}
mt_named <- data.frame(mtcars, name = row.names(mtcars))

p <- ggplot(mt_named, aes(mpg, reorder(name, mpg))) + 
  geom_point(size = 2.5, color = "#0072B2") +
  geom_text(aes(x = mpg + .5, label = name), hjust = 0, size = 3) +
  scale_y_discrete(name = NULL, breaks = NULL, expand = c(0, 1)) +
  theme_bw() +
  theme(panel.border = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line.x = element_line(),
        plot.margin = margin(6, 45, 6, 6))
p
```
screen shot 2018-04-27 at 12 59 24 pm

The top-most text label is cut off because it extends beyond the plot area. To make it fit, I could extend the axis range, but that would extend the axis line as well. It looks more elegant to switch of clipping and have the text label extend outside of the plot area:
```{r}
g <- ggplotGrob(p)
index <- grep("panel", g$layout$name)
g$layout$clip[index] = "off"
grid::grid.newpage()
grid::grid.draw(g)
```
screen shot 2018-04-27 at 1 00 14 pm

I would like to propose to address this by making clipping a feature of the coordinate system. For example, the above plot could be generated by something like this:
```{r}
p + coord_cartesian(clip = "off")
```
This feature requires very minimal changes to the coord code and the facet layouting code. I'll see if I can put together a pull request implementing this.

### PR DESCRIPTION
**Pull request by clauswilke**
This pull request makes it possible to switch off clipping via a setting in the coordinate functions. Examples follow.

First a regular plot with standard clipping:
```{r}
mt_named <- data.frame(mtcars, name = row.names(mtcars))
p <- ggplot(mt_named, aes(mpg, reorder(name, mpg))) + 
  geom_point(size = 2.5, color = "#0072B2") +
  geom_text(aes(x = mpg + .5, label = name), hjust = 0, size = 3) +
  scale_y_discrete(name = NULL, breaks = NULL, expand = c(0, 1)) +
  theme_bw() +
  theme(panel.border = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line.x = element_line(),
        plot.margin = margin(6, 45, 6, 6))
p
```
screen shot 2018-04-27 at 1 59 19 pm

Clipping turned off:
```{r}
p + coord_cartesian(clip = "off")
```
screen shot 2018-04-27 at 1 59 26 pm

Fixed coordinate system:
```{r}
p + coord_fixed(clip = "off")
```
screen shot 2018-04-27 at 1 59 34 pm

This works also with `facet_wrap()` and `facet_grid()`:
```{r}
p + coord_cartesian(clip = "off") +
  facet_wrap(~carb)
```
screen shot 2018-04-27 at 1 59 51 pm

```{r}
p + coord_cartesian(clip = "off") +
  facet_grid(gear~carb)
```
screen shot 2018-04-27 at 2 00 01 pm

Nothing is changed from current ggplot2 as long as we don't set the `clip` option: 
```{r}
p + coord_cartesian() + facet_grid(gear~carb)
```
screen shot 2018-04-27 at 2 00 37 pm

### CODE REVIEW

File: NEWS [hunk 1]
```diff
@@ -66,6 +66,9 @@ BUG FIXES AND MINOR IMPROVEMENTS
   quoted calls/names. `aes_string()` uses names `x` and `y` for first
   two unnamed arguments.
 
+* Clipping to the plot panel is now configurable, through a `clip` argument
```
Comment (hadley) [inline NEWS]: Can you please reference issue number and your user name?

**Comment by hadley** [comment]
I'll wait to merge this until visual testing is easy again - I think we need one test here to prevent regressions.

**Comment by thomasp85** [comment]
I just want to chime in that it is very important to fully document what is happening and what to expect - if you set the limits in the coordinate system and set `clip = "off"` tha data outside the panel will be all over your margins which could probably freak someone out a bit...

**Comment by clauswilke** [comment]
I have expanded the documentation as requested. Note that the parameter `default` is not currently properly documented. It says:

> `default` Is this the default coordinate system? If `FALSE`

I'm happy to complete the sentence if somebody could tell me what it should be.

I have also added a few visual tests for the clip option. Two example tests:

```{r}
df.in <- data.frame(label = c("inside", "inside", "inside", "inside"),
                    x = c(0, 1, 0.5, 0.5),
                    y = c(0.5, 0.5, 0, 1),
                    angle = c(90, 270, 0, 0),
                    hjust = c(0.5, 0.5, 0.5, 0.5),
                    vjust = c(1.1, 1.1, -0.1, 1.1))

df.out <- data.frame(label = c("outside", "outside", "outside", "outside"),
                     x = c(0, 1, 0.5, 0.5),
                     y = c(0.5, 0.5, 0, 1),
                     angle = c(90, 270, 0, 0),
                     hjust = c(0.5, 0.5, 0.5, 0.5),
                     vjust = c(-0.1, -0.1, 1.1, -0.1))


p <- ggplot(mapping = aes(x, y, label = label, angle = angle, hjust = hjust, vjust = vjust)) + 
  geom_text(data = df.in) +
  geom_text(data = df.out) +
  scale_x_continuous(breaks = NULL, name = NULL) +
  scale_y_continuous(breaks = NULL, name = NULL) +
  theme(plot.margin = margin(20, 20, 20, 20),
        panel.spacing = grid::unit(10, "pt"))

p + coord_cartesian(xlim = c(0, 1), ylim = c(0, 1), expand = FALSE)
```
screen shot 2018-05-01 at 1 42 40 pm


```{r}
p + coord_cartesian(xlim = c(0, 1), ylim = c(0, 1), expand = FALSE, clip = "off")
```
screen shot 2018-05-01 at 1 42 46 pm

**Comment by hadley** [comment]
Do you think you could reduce the number of visual tests? They're quite expensive, and I think for most of them you could get away with verifying that `clip` is set correctly in the coord object.

**Comment by clauswilke** [comment]
I have reduced the visual tests to two, one with and one without clipping, and have added a couple of non-visual tests for the clip setting. I have also fixed a documentation bug that prevented CRAN check from running without error. It should all work now.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**